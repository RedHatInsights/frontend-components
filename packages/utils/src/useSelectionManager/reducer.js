import isObject from 'lodash/isObject';
import uniq from 'lodash/uniq';

const selectionGroup = (action) => action.group || 'default';

export const init = (withGroups) => (preselected) => withGroups ? preselected || {} : { default: preselected || [] };

const cleanEmpty = (state) => {
  const newState = state;
  Object.entries(state).forEach(([key, value]) => {
    if (value === undefined) {
      delete newState[key];
    }
  });
  return newState;
};

const set = (state = {}, action) => {
  const group = selectionGroup(action);

  return cleanEmpty({
    ...state,
    [group]: action.items?.length > 0 || isObject(action.items) ? action.items : undefined,
  });
};

const select = (state = {}, action) => {
  const group = selectionGroup(action);
  return cleanEmpty({
    ...state,
    [group]: action.reset ? action?.items : uniq([action?.item, ...(state[group] || [])]),
  });
};

const deselect = (state = {}, action) => {
  const group = selectionGroup(action);
  const items = (state[group] || []).filter((selectedItem) => selectedItem !== action?.item);

  return cleanEmpty({
    ...state,
    [group]: items.length === 0 ? undefined : items,
  });
};

const toggle = (state, action) => {
  const group = selectionGroup(action);
  return (state[group] || []).includes(action.item) ? deselect(state, action) : select(state, action);
};

const reset = (state, action) => init(!state.hasOwnProperty('default'))(action?.preselected);
const clear = (state) => init(!state.hasOwnProperty('default'))();

export default (state, action) =>
  ({
    set,
    select,
    deselect,
    toggle,
    reset,
    clear,
  }[action.type](state, action));
