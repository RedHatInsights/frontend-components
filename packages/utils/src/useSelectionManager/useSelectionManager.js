import { useReducer } from 'react';
import reducer, { init as initReducer } from './reducer';

const useSelectionManager = (preselected, options = {}) => {
  const { withGroups = false } = options;
  const [selection, dispatch] = useReducer(reducer, preselected, initReducer(withGroups));

  const set = (items, group) => dispatch({ type: 'set', group, items });

  const select = (item, group, useSet = false) => (useSet ? set(item, group) : dispatch({ type: 'select', group, item }));

  const deselect = (item, group, useSet = false) => (useSet ? set(item, group) : dispatch({ type: 'deselect', group, item }));

  const toggle = (item, group) => dispatch({ type: 'toggle', group, item });

  const reset = () => dispatch({ type: 'reset', preselected });

  const clear = () => dispatch({ type: 'clear' });

  return {
    set,
    select,
    deselect,
    toggle,
    reset,
    clear,
    selection: withGroups ? selection : selection.default,
  };
};

export default useSelectionManager;
