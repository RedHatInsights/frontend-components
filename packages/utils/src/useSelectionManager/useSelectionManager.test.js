import { act, renderHook } from '@testing-library/react-hooks';
import useSelectionManager from './useSelectionManager';

describe('useSelectionManager', () => {
  it('returns an object with function to manage selections', () => {
    const { result } = renderHook(() => useSelectionManager());
    expect(result).toMatchSnapshot();
  });

  describe('withGroups: false', () => {
    const defaultArguments = [[1, 2, 3, 4]];

    it('returns an object with function to manage selections wihtout groups', () => {
      const { result } = renderHook(() => useSelectionManager(...defaultArguments));
      expect(result.current.selection).toEqual(defaultArguments[0]);
      expect(result.current).toMatchSnapshot();
    });

    it('adds an item from the selection when calling select', () => {
      const { result } = renderHook(() => useSelectionManager(...defaultArguments));

      act(() => {
        result.current.select(42);
      });

      expect(result.current).toMatchSnapshot();
    });

    it('removes an item from the selection when calling deselect', () => {
      const { result } = renderHook(() => useSelectionManager(...defaultArguments));

      act(() => {
        result.current.deselect(3);
      });

      expect(result.current).toMatchSnapshot();
    });

    it('sets items for a selection when calling set', () => {
      const { result } = renderHook(() => useSelectionManager(...defaultArguments));

      act(() => {
        result.current.set([0, 9, 8, 45, 3]);
      });

      expect(result.current).toMatchSnapshot();
    });
  });

  describe('withGroups true', () => {
    const defaultArguments = [{ group1: [1, 2, 3, 4], group2: [12, 23, 34, 45] }, { withGroups: true }];

    it('returns an object with function to manage selections with groups', () => {
      const { result } = renderHook(() => useSelectionManager(...defaultArguments));
      expect(result.current).toMatchSnapshot();
    });

    it('adds an item from the selection when calling select', () => {
      const { result } = renderHook(() => useSelectionManager(...defaultArguments));

      act(() => {
        result.current.select(42, 'group2');
      });

      expect(result.current).toMatchSnapshot();
    });

    it('removes an item from the selection when calling deselect', () => {
      const { result } = renderHook(() => useSelectionManager(...defaultArguments));

      act(() => {
        result.current.deselect(2, 'group1');
      });

      expect(result.current).toMatchSnapshot();
    });

    it('sets items for a selection when calling set', () => {
      const { result } = renderHook(() => useSelectionManager(...defaultArguments));

      act(() => {
        result.current.set([0, 9, 8, 45, 3], 'group1');
      });

      expect(result.current).toMatchSnapshot();
    });
  });
});
