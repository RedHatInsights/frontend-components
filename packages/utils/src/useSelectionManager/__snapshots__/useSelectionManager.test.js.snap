// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSelectionManager returns an object with function to manage selections 1`] = `
Object {
  "all": Array [
    Object {
      "clear": [Function],
      "deselect": [Function],
      "reset": [Function],
      "select": [Function],
      "selection": Array [],
      "set": [Function],
      "toggle": [Function],
    },
  ],
  "current": Object {
    "clear": [Function],
    "deselect": [Function],
    "reset": [Function],
    "select": [Function],
    "selection": Array [],
    "set": [Function],
    "toggle": [Function],
  },
  "error": undefined,
}
`;

exports[`useSelectionManager withGroups true adds an item from the selection when calling select 1`] = `
Object {
  "clear": [Function],
  "deselect": [Function],
  "reset": [Function],
  "select": [Function],
  "selection": Object {
    "group1": Array [
      1,
      2,
      3,
      4,
    ],
    "group2": Array [
      42,
      12,
      23,
      34,
      45,
    ],
  },
  "set": [Function],
  "toggle": [Function],
}
`;

exports[`useSelectionManager withGroups true removes an item from the selection when calling deselect 1`] = `
Object {
  "clear": [Function],
  "deselect": [Function],
  "reset": [Function],
  "select": [Function],
  "selection": Object {
    "group1": Array [
      1,
      3,
      4,
    ],
    "group2": Array [
      12,
      23,
      34,
      45,
    ],
  },
  "set": [Function],
  "toggle": [Function],
}
`;

exports[`useSelectionManager withGroups true returns an object with function to manage selections with groups 1`] = `
Object {
  "clear": [Function],
  "deselect": [Function],
  "reset": [Function],
  "select": [Function],
  "selection": Object {
    "group1": Array [
      1,
      2,
      3,
      4,
    ],
    "group2": Array [
      12,
      23,
      34,
      45,
    ],
  },
  "set": [Function],
  "toggle": [Function],
}
`;

exports[`useSelectionManager withGroups true sets items for a selection when calling set 1`] = `
Object {
  "clear": [Function],
  "deselect": [Function],
  "reset": [Function],
  "select": [Function],
  "selection": Object {
    "group1": Array [
      0,
      9,
      8,
      45,
      3,
    ],
    "group2": Array [
      12,
      23,
      34,
      45,
    ],
  },
  "set": [Function],
  "toggle": [Function],
}
`;

exports[`useSelectionManager withGroups: false adds an item from the selection when calling select 1`] = `
Object {
  "clear": [Function],
  "deselect": [Function],
  "reset": [Function],
  "select": [Function],
  "selection": Array [
    42,
    1,
    2,
    3,
    4,
  ],
  "set": [Function],
  "toggle": [Function],
}
`;

exports[`useSelectionManager withGroups: false removes an item from the selection when calling deselect 1`] = `
Object {
  "clear": [Function],
  "deselect": [Function],
  "reset": [Function],
  "select": [Function],
  "selection": Array [
    1,
    2,
    4,
  ],
  "set": [Function],
  "toggle": [Function],
}
`;

exports[`useSelectionManager withGroups: false returns an object with function to manage selections wihtout groups 1`] = `
Object {
  "clear": [Function],
  "deselect": [Function],
  "reset": [Function],
  "select": [Function],
  "selection": Array [
    1,
    2,
    3,
    4,
  ],
  "set": [Function],
  "toggle": [Function],
}
`;

exports[`useSelectionManager withGroups: false sets items for a selection when calling set 1`] = `
Object {
  "clear": [Function],
  "deselect": [Function],
  "reset": [Function],
  "select": [Function],
  "selection": Array [
    0,
    9,
    8,
    45,
    3,
  ],
  "set": [Function],
  "toggle": [Function],
}
`;
